### PR TITLE
simulated mouse fixes, rightclick, dblclick

### LINKED ISSUE #53: Need a "rightclick" command
I need the ability to right click an element within Cypress so that it triggers the `contextmenu` event and all that entails.

### LINKED ISSUE #116: cy.dblclick() should accept same options as cy.click()
`cy.dblclick()` currently will serially double click all of the current subjects. `cy.dblclick()` needs to accept options:
- `multiple`
- `force`
- `interval`

It also needs to accept `position` and `x`,`y` coordinate arguments.

### LINKED ISSUE #1016: clicking on an openlayers canvas doesn't trigger any events
Hi,

When i have a test like this:

```js
it('should draw little circle on a map', function () {
	cy.visit('http://openlayers.org/en/latest/examples/draw-features.html');
	cy.get('canvas').click();
	cy.get('canvas').click(100, 100);
});
```

it doesn't trigger any events (the openlayers example should draw dots on the map)
Perhaps this is because the mousemove and mouseover events are not triggered... i don't know the inner workings of openlayers on this... (although i do know they do stuff a little different)

the problem we now have i can't test any geo applications based on openlayers (which are most of them)

### LINKED ISSUE #1847: Cypress should fire 'mouseover' event on .click()
### Current behavior:

Unable to click on the first suggestion from google place auto complete

Video: https://drive.google.com/open?id=11dxli8EzoozjZ0LfAmDT6IJKHWKtlNKy

### Desired behavior:

Can click on the first suggestion

### Steps to reproduce:

The following script would fail. For some reason click wouldn't pick the item here. I've tried selenium and the click worked for me.

```js
cy
  .visit('https://developers.google.com/maps/documentation/javascript/examples/full/places-autocomplete')
  .get('#pac-input')
  .should('be.visible')
  .wait(1000)
  .type('22 Princes Highway, Darlington NSW, Australi')
  .get('.pac-item')
  .first()
  .should('be.visible')
  .click()
  .get('#pac-input')
  .then((result) => {
      expect(result.val()).to.eq('22 Princes Highway, Darlington NSW, Australia');
  });
```

### Versions

Cypress 3.0.1
Mac 10.13.4
Electron 59

### LINKED ISSUE #2045: Conversion to onPointerDown
### Current behavior:
We are converting our app to React 16.4 which introduced Pointer Events API. However, items that we convert from onClick/onMouseDown-> onPointerDown break our Cypress tests. Everything works normally on our dev environment, and we have a PointerEvents polyfill as well (cypress' Chrome version supports the API natively).

Duplicating our `onPointerDown={someAction}` to `onPointerDown={someAction} onMouseDown={someAction}` fixes the tests but adds duplication. 

### Desired behavior:
Cypress should be able to handle onPointerDown events without code duplication

### Versions
Cypress 3.0.1

### LINKED ISSUE #2521: No mousemove events before clicks
### Current behavior:

When I `cy.get('...').click();` no `mousemove` events are triggered, so every mousemove related scripts are broken.

Here an example showing the GitHub's milestone panel broke because of the missing mousemove events. I run the same test with Puppeteer and TestCafè too to understand what is the common behavior in such scenario

and here the log related to the three frameworks

TestCafè
```
mouseover <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label="Set milestone" aria-haspopup="menu" data-hotkey="m">...</summary>
mousemove <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label="Set milestone" aria-haspopup="menu" data-hotkey="m">...</summary>
mousemove <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label="Set milestone" aria-haspopup="menu" data-hotkey="m">...</summary>
mousedown <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label="Set milestone" aria-haspopup="menu" data-hotkey="m">...</summary>
mouseup <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label="Set milestone" aria-haspopup="menu" data-hotkey="m">...</summary>
click <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label="Set milestone" aria-haspopup="menu" data-hotkey="m">...</summary>
```

Puppeteer
```
mouseover <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label=​"Set milestone" aria-haspopup=​"menu" data-hotkey=​"m">​…​</summary>​
mousemove <summary class=​"text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label=​"Set milestone" aria-haspopup=​"menu" data-hotkey=​"m">​…​</summary>​
mousedown <summary class=​"text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label=​"Set milestone" aria-haspopup=​"menu" data-hotkey=​"m">​…​</summary>​
mouseup <summary class=​"text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label=​"Set milestone" aria-haspopup=​"menu" data-hotkey=​"m">​…​</summary>​
click <summary class=​"text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label=​"Set milestone" aria-haspopup=​"menu" data-hotkey=​"m">​…​</summary>​
```

Cypress
```
mousedown <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label="Set milestone" aria-haspopup="menu" data-hotkey="m">...</summary>
mouseup <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label="Set milestone" aria-haspopup="menu" data-hotkey="m">...</summary>
click <summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle" aria-label="Set milestone" aria-haspopup="menu" data-hotkey="m">...</summary>
```

### Desired behavior:

I know that I can simulate the mousemove event and if this is the exactly expected behavior... sorry if I bored you...

### Steps to reproduce:

In [this repo](https://github.com/NoriSte/stackoverflow-52383438-cypress-issue) you can find the most complete version of my tests but I prepared even an [easiest one](https://github.com/NoriSte/cypress-mousemove-issue) without any useless stuff.

### Versions

Cypress 3.1.0 on Mac OSX 10.13.6

### LINKED ISSUE #2768: Map not responding to mouse commands
### Current behavior:

No response from the map in application under test to mouse commands in Cypress test script. 

### Desired behavior:

Looking for a way to update test script so that mouse commands work with the map, or updates to Cypress.

### Steps to reproduce:

This test script has four different attempts to use mouse commands on the map in this application, none of them are working.
```js
describe('mouse test', function () {
	before('open application', function () {
		cy.visit('https://master-branch-opensphere-ngageoint.surge.sh/?tips=false')
		cy.get('.webgl-canvas', {timeout: 12000}) // map loaded
		cy.get('body').type('+++++++++++++++++++++++++') // zoom in on the map
		cy.get('.close > span').click() // close layers widget
	})

	it('open map context menu', function () {
		cy.get('.webgl-canvas').trigger('contextmenu')  // attempt to open the context menu
		cy.get('#menu') // check for menu
	})

	it('zoom on double click', function () {
		cy.get('.zoom-text').should('have.text', 'Zoom: 6.2')// verify current zoom level
		cy.get('.webgl-canvas')
			.dblclick()
			.dblclick()
			.dblclick()
			.dblclick()
			.dblclick()
			.dblclick()
			.dblclick()
			.dblclick()
		cy.get('.zoom-text').should('not.have.text', 'Zoom: 6.2')// verify current zoom level
	})


	it('load query area (query button)', function () {
		cy.get('[title="Draws a box on the map for queries, zoom, and selection"]').click() // activate query tool
		cy.get('[class="btn btn-primary on active"]') // verify query tool active

		cy.get('body').as('map')
		cy.get('@map') // attempt to draw a query area
			.trigger('mousedown', "center", {force:true})
			.trigger('mousemove', 400, 400, {force:true})
			.trigger('mouseup', 400, 400, {force:true})

		cy.get('#menu') // check for menu
	})

	it('load query area (keyboard and mouse)', function () {
		cy.get('body').as('map')
		cy.get('@map') // attempt to draw a query area
			.type('{ctrl}', {release:false,force:true}).click(400,400)
			.type('{ctrl}', {release:true,force:true}).click(600,600)
			
		cy.get('#menu') // check for menu
	})
})
```
### Versions

Cypress 3.1.1
Chrome 70
Windows 10 / macOS 10.12.6

### LINKED ISSUE #2928: click not firing mouseover event
### Current behavior:

I trigger some react-contextify menu with `.trigger('contextmenu')`
My menu is displaying OK

image

Then I select the div element where an onClick event is attached.
`cy.get('.react-contexify__item').debug().click();`

I tried to use also "force: true" as an option.

But my action is not triggered. Nothing happens.

With debug() option activated, I trigger manually the click() event and my Material UI dialog is displayed.

``` javascript
------------------------ Debug Info ------------------------
cypress_runner.js:63678 Command Name:     get
cypress_runner.js:63678 Command Args:     [".react-contexify__item"]
cypress_runner.js:63678 Current Subject:  jQuery.fn.init [div.react-contexify__item, prevObject: jQuery.fn.init(1), context: document, selector: ".react-contexify__item"]
subject.click()
MegaTrendPage.jsx:316 hello!
jQuery.fn.init [div.react-contexify__item, prevObject: jQuery.fn.init(1), context: document, selector: ".react-contexify__item"]
```
Source for div menu is like this: 

``` javascript
  return (
      <div
        className={cssClasses}
        style={style}
        onClick={this.handleClick}
        role="presentation"
      >
        <div className={styles.itemContent}>{children}</div>
      </div>
    );
```

### Desired behavior:

I trigger click event and dialog is shown...

### Versions

Cypress Version used : 3.1.3
Chrome 70.0.3538.110 
React 16.6.3
react-contexify : 3.0.3

### LINKED ISSUE #2956: [moved] Simulated click fixes
For tracking issues with simulated click related to native events

Edit: migrated to PR comment of #3030

### LINKED ISSUE #2957: click not firing pointer events
- [x] We should fire pointerdown + pointerup during a click at least

### browser support:

Feature | Chrome | Firefox (Gecko) | Internet Explorer | Opera | Safari (WebKit)
-- | -- | -- | -- | -- | --
Basic support | 55.0 | 59 (59) | 10 | 42 | No support

### All pointer events:
pointerover
pointerenter
pointerdown
pointermove
pointerup
pointercancel
pointerout
pointerleave
gotpointercapture
lostpointercapture

### LINKED ISSUE #3187: trigger('mousemove', ...) yields undefined values for screenX and screenY in event handler
### Current behavior:

Triggering `mousemove` via `.trigger` correctly fills in `pageX`, `pageY`, `clientX`, and `clientY`. However, Cypress fails to fill in the `screenX` and `screenY` values.
This can cause erroneous behavior if the application depends on them.

### Desired behavior:

`screenX` and `screenY` should probably just have the same values as `clientX` and `clientY`.

### Versions

3.1.4

### LINKED ISSUE #3224: Cypress's dblclick does not function the same way as a user double clicking
### Current behavior:
Firing Cypress's `dblclick()` does not behave the same way as a user double clicking an element.  Specifically, if there is no special meaning for a double click, then when a user double clicks, the first `click` event triggers a `click` action.  However, with `dblclick()`, if nothing is defined specifically for a double click, then the element is put in focus, but the action associated with a single click does not occur. See test example below.  

### Desired behavior:
Firing Cypress's `dblclick()` behaves the same way as a user double clicking an element.  I would expect that in both the examples below Cypress would navigate to the "mission" page, because a user double clicking on a link would be interpreted as a single `click` event if no special meaning is defined for the double click.

### Steps to reproduce:
```
it('Example of successful single click', () => {
  cy.visit('https://www.google.com');
  cy.contains('About').click();
  cy.contains('Our mission is to');
});

it('Example of unsuccessful double click', () => {
  cy.visit('https://www.google.com');
  cy.contains('About').dblclick();
  cy.contains('Our mission is to');
});
```

### Versions
Windows 10
Cypress 3.1.4
Chrome 71

### LINKED ISSUE #3798: dblclick() doesn't work as click()
### Current behavior:

> Error:       CypressError: cy.dblclick() failed because this element is not visible:
> 
> <span class="pax-add pax-enabled" id="hrefIncAdt">+</span>
> 
> This element '<span#hrefIncAdt.pax-add.pax-enabled>' is not visible because its content is being clipped by one of its parent elements, which has a CSS property of overflow: 'hidden', 'scroll' or 'auto'
> 
> Fix this problem, or use {force: true} to disable error checking.

### Desired behavior:

- The double click function gets NO error as click function works.

### Steps to reproduce: (app code and test code)
Having the code to demonstrate the issue as below

```js
context('Drop-down with navigation', () =>{
        it.only('Check drop-down with navigation', ()=>{
            cy.visit('https://www.spicejet.com/')
            cy.get('.adult-infant-child').within(() => {
                cy.get('#divpaxinfo').click()
                    // choose number of adult
                    cy.get('#divAdult #hrefIncAdt').dblclick()
                    // choose number of child
                    cy.get('#divChild #hrefIncChd').click()

            })
        })
    })
```

1. Run the code above => get error

2. Work around:
Replace the line 
`cy.get('#divAdult #hrefIncAdt').dblclick()`
by
`cy.get('#divAdult #hrefIncAdt').click()`

### Versions

```json
  "devDependencies": {
    "cypress": "^3.1.5",
    "cypress-xpath": "^1.3.0",
    "eslint-plugin-cypress": "^2.2.1"
  }
```
node version: v9.3.0
OS: MAC OS

### LINKED ISSUE #4347: Cypress click coordinates incorrect when element inside Iframe
Cypress calculates coords for the click assuming the element is in the topmost frame, which causes coords to be off:

I clicked the `hashchange`, which was at the top of an iframe:
image

So Cypress thought I clicked the top of my AUT frame:
image

this will be fixed in #3030

### PR DESCRIPTION
**Pull request by kuceb**
### mouse events
mouse events added in this PR:
```diff
+ pointerup
+ pointerdown
+ pointermove
+ pointerover
+ pointerenter
+ pointerleave
+ pointerout
mouseup
mousedown
+ mousemove
+ mouseover
+ mouseenter
+ mouseleave
+ mouseout
click
dblclick
+ contextmenu
```
- [x] send mouseover event during click commands 
fix #2928 
fix #1847
fix #1016
- [x] send pointer events (`pointerdown`, `pointerup`, `pointermove`) during click commands 
fix #2957
fix #2045
- [x] send mousemove events during click commands 
fix #2521 
- [x] add mouseout events

- [x] fix incorrect click coordinates inside [same domain] iframe
fix #4347

### right click, double click
- [x] allow `dblclick` command to accept same options + perform actions similar to `click`
fix #116 
fix #3798
fix #2768
fix #3224
- [x] add "rightclick" command
fix #53 

### trigger
- [x] add `screenX`/`screenY` properties to events sent with `.trigger`
fix #3187 
  - [x] tests

#### tasks
- [x] pointer events
- [x] console log shows all events
- [x] fix dblclick
- [x] right click
- [x] ~send `selectionchange` event on `document`~ _browser already does this even when out of focus_
- [x] add test for clicking inside iframe
- [x] fix coords incorrect when clicking inside iframe
- [x] waiting on fix for #4396 
- [x] typedefs
- [x] fix messed up console table of events
- [x] add test for cy.click checking checkboxes
- [x] add comment for element-moving-during-click algorithm
- [x] refactor action commands to reuse click logic
- [x] add `fromAutWindow` positional coords for red hitbox
- [x] add test for iframe red hitbox coords
 
#### non-blocking tasks
- [ ] write new doc for .rightclick() command https://github.com/cypress-io/cypress-documentation/issues/2102
- [ ] update document events simulated in `.click()` https://docs.cypress.io/api/commands/click.html#Events https://github.com/cypress-io/cypress-documentation/issues/2173
- [ ] update `.dblclick()` doc `options` to match `.click()` options. https://docs.cypress.io/api/commands/dblclick.html#Arguments
- [ ] kitchensink update https://github.com/cypress-io/cypress-example-kitchensink/pull/298

### Guide for review

- review console messages during `cy.click, cy.dblclick, and cy.rightclick`
- make sure tests cover all use cases of `dblclick` and `rightclick`
- let me know if I missed an issue this fixes

#### API:
- add `cy.rightclick()`

#### typedefs changes:
- add `cy.rightclick()` typedef

#### UX:
- console messages when pinning `click` command has more information:
image
warning messaging in table:
image
no mouse move events needed:
image

### CODE REVIEW

**Comment by Bkucera** [comment]
WIP going back on for dblclick and rightclick changes

**Comment by jennifer-shehane** [comment]
Changed the base to `develop`, is this still WIP since there's a checkbox not checked?

**Comment by Bkucera** [comment]
It's in a reviewable state, but I do still have one task to do

**Comment by brian-mann** [comment]
I see some cleanup but overall this is excellent.

**Review by brian-mann — DISMISSED** [review]
- these are mostly notes, but adding them now. some of which i will internally handle - so don't jump into these yet.
- i'm currently going through each of the issues and finding the tests and running them to understand and gauge the solutions we implemented.
- also i'm looking for our collective notes on the algorithm changes so we can surface those and them not get lost
- we talked internally about using `while` loops vs tail-call optimized recursions. no change needed but some research needs to be done to see which pattern we should employ. there are places that we traverse the dom using recursion which may not be tail-call optimized and could blow out the stack. however research suggests recursion is more performant that while loops. probably is easier to convert those recursive calls to `while` loops, but ideally we could look into this further and then decide on a general strategy of how/when to use one vs the other.

File: packages/driver/test/cypress/integration/commands/actions/click_spec.js [hunk 1]
```diff
@@ -257,25 +294,37 @@ describe('src/cy/commands/actions/click', () => {
...
-          return events.push(event)
-        })
-      })
+      attachMouseClickListeners({ $btn })
 
-      cy.get('#button').click().then(() => {
-        expect(events).to.deep.eq(['mouseup', 'click'])
+      cy.get('#button').click().should('not.have.focus')
+
+      cy.getAll('$btn', 'pointerdown mousedown pointerup mouseup click').each(shouldBeCalled)
```
Comment (brian-mann) [inline packages/driver/test/cypress/integration/commands/actions/click_spec.js]: need to think about a potentially better way to do this. not necessarily bad or wrong just need to discuss

File: packages/driver/src/cy/mouse.js [hunk 2]
```diff
@@ -1,143 +1,653 @@
 const $dom = require('../dom')
+const $elements = require('../dom/elements')
+const $ = require('jquery')
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cy/mouse.js]: we have not. there is a plugin but it has strange bugs sometimes where it moves all your requires to the bottom of the file

File: packages/driver/src/cy/mouse.js [hunk 3]
```diff
@@ -1,143 +1,653 @@
...
-        clientY: fromViewport.y,
-        buttons: 0,
+      const defaultOptions = mouse._getDefaultMouseOptions(x, y, win)
+
+      const pointerEvtOptions = _.extend({}, defaultOptions, {
+        button: 0,
+        which: 1,
+        buttons: 1,
+        detail: 0,
+        pressure: 0.5,
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cy/mouse.js]: yes, added comment to clarify

File: packages/driver/src/cy/mouse.js [hunk 4]
```diff
@@ -1,143 +1,653 @@
...
 
-module.exports = { create }
+const { stopPropagation } = window.MouseEvent.prototype
+
+const sendEvent = (evtName, el, evtOptions, bubbles = false, cancelable = false, constructor) => {
+  evtOptions = _.extend({}, evtOptions, { bubbles, cancelable })
+  const _eventModifiers = $Keyboard.fromModifierEventOptions(evtOptions)
+  const modifiers = $Keyboard.modifiersToString(_eventModifiers)
+
+  const evt = new constructor(evtName, _.extend({}, evtOptions, { bubbles, cancelable }))
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cy/mouse.js]: :+1:

File: packages/driver/src/cy/mouse.js [hunk 5]
```diff
@@ -1,143 +1,653 @@
...
+
+const sendPointerEvent = (el, evtOptions, evtName, bubbles = false, cancelable = false) => {
+  const constructor = el.ownerDocument.defaultView.PointerEvent
+
+  return sendEvent(evtName, el, evtOptions, bubbles, cancelable, constructor)
+}
+const sendMouseEvent = (el, evtOptions, evtName, bubbles = false, cancelable = false) => {
+  // IE doesn't have event constructors, so you should use document.createEvent('mouseevent')
+  // https://dom.spec.whatwg.org/#dom-document-createevent
+  const constructor = el.ownerDocument.defaultView.MouseEvent
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cy/mouse.js]: :+1:

File: packages/driver/src/cy/mouse.js [hunk 6]
```diff
@@ -1,143 +1,653 @@
...
+     */
+    getElAtCoordsOrForce ({ x, y, doc }, forceEl) {
+      if (forceEl) {
+        return forceEl
       }
 
-      const modifiers = keyboard.getActiveModifiersArray()
+      const el = doc.elementFromPoint(x, y)
+
+      // mouse._mouseMoveEvents(el, { x, y })
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cy/mouse.js]: :+1:

File: packages/driver/src/cy/mouse.js [hunk 7]
```diff
@@ -1,143 +1,653 @@
 const $dom = require('../dom')
+const $elements = require('../dom/elements')
+const $ = require('jquery')
+const _ = require('lodash')
+const $Keyboard = require('./keyboard')
+const $selection = require('../dom/selection')
+const debug = require('debug')('driver:mouse')
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cy/mouse.js]: :+1:

File: packages/driver/src/cy/keyboard.js [hunk 8]
```diff
@@ -37,6 +38,34 @@ const initialModifiers = {
...
+    ctrl: eventOptions.ctrlKey,
+    meta: eventOptions.metaKey,
+    shift: eventOptions.shiftKey,
+
+  }, Boolean)
+}
+
+const modifiersToString = (modifiers) => {
+  return _.keys(
+    _.pickBy(modifiers, (val) => {
```
[the remark this answers is not in the document]
Comment (Bkucera) [reply packages/driver/src/cy/keyboard.js]: :+1:

File: packages/driver/test/cypress/integration/commands/actions/click_spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (brian-mann) [reply packages/driver/test/cypress/integration/commands/actions/click_spec.js]: This was an old comment, disregard

**Review by brian-mann — CHANGES_REQUESTED** [review]
Can you write out the event firing algorithm including how we handle things like when the element moves in response to an event? 

It will take some legwork, and it may possibly be better diagramed as a flowchart (rather than nested bullet points) - i don't necessarily care, i just want it documented somewhere. I would prefer if its documented in the code itself so it doesn't become stale - and new engineers will be able to pick it up. I'd also accept a link to an external document in the code if necessary. We can also use flowchart.js to accomplish both - where the algorithm is written in code in notes but a physical flowchart asset is generated from the code. That way it never gets stale.

as you described they are incoherent. We need to build institutional knowledge around these algorithms and make them very transparent. Jennifer created one awhile ago for the visibility algorithm - she'll link you here in a sec.

**Comment by jennifer-shehane** [comment]
maybe flowchart in code similar visibility algo 

nC0TrUIVMOLJObyg-F2826 copy

**Comment by Bkucera** [comment]
@jennifer-shehane ok, do we have the flowchart-as-code stuff documented anywhere

**Review by brian-mann — CHANGES_REQUESTED** [review]
update test for iframe coordinates to actually have coords
- fromAutWindow
- remove the $elements.getFromDocCoords(x, y, win)
- write a proper test for the iframe coordinates specifically for drawing the red hitbox
- remove docCoords

**Pull request merged**